### PR TITLE
refactor/improve CRD publishing e2e tests in an HA setup

### PR DESCRIPTION
**Pull request by p0lyn0mial**

### CODE REVIEW

File: vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go [hunk 1]
```diff
@@ -792,3 +776,186 @@ properties:
...
+	name := "wait-for-crd-definition"
+	return &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name + "-" + string(uuid.NewUUID()),
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:    name,
+					Image:   "alpine",
```
Comment (p0lyn0mial) [inline vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go:847]: Is it okay to use an image like this one?
I couldn't find one (`imageutils.GetE2EImage`) that would have or allow me to install `curl, jq`

Comment (smarterclayton) [reply vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go:847]: No unfortuntaely ti is not.  You must use an image defined in the upstream utils/image package.

Comment (smarterclayton) [reply vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go:847]: If no existing upstream image supports jq you can't use it in your test until you get an image added, THEN you need to get that added upstream, THEN it needs to be part of upstream image publication.

Comment (p0lyn0mial) [reply vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go:847]: oh, okay, thanks for the info.

File: vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go [hunk 2]
```diff
@@ -792,3 +786,187 @@ properties:
...
+       PREV_HTTP_BODY=""
+       PREV_ETAG=""
+
+       while [ ${SEEN} -lt ${SERVERS_LEN} ]
+       do
+         for server in ${SERVERS}
+         do
+           URL="https://$server/openapi/v2"
+           echo "$(date):downloading the OpenAPI spec from ${URL}, prev_etag=${PREV_ETAG}, seen=${SEEN}"
+           HTTP_RESPONSE=$(curl -k -s -w "HTTPSTATUS:%{http_code}" -H "If-None-Match:\"${PREV_ETAG}\"" -H "Authorization: Bearer $TOKEN" --dump-header rsp-header.json ${URL})
```
Comment (p0lyn0mial) [inline vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go:806]: TODO(me): fill in the User-Agent with the given namespace

File: vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go (hunk 1 above)
Comment (p0lyn0mial) [reply vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go:848]: @smarterclayton would you know how to add a new image?

File: vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go [hunk 3]
```diff
@@ -520,9 +519,23 @@ func setupCRDAndVerifySchema(f *framework.Framework, schema, expect []byte, grou
 	if err != nil {
 		return nil, fmt.Errorf("failed to create CRD: %v", err)
 	}
+	return crd, nil
+}
+
+func setupCRDAndVerifySchema(f *framework.Framework, schema []byte, groupSuffix string, versions ...string) (*crd.TestCrd, error) {
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go:525]: what does verify mean here?

File: vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go [hunk 4]
```diff
@@ -792,3 +786,187 @@ properties:
...
+			framework.Logf("error pulling logs: %v", err)
+			return err
+		}
+		framework.Logf("pooled logs:\n%v", logs)
+	}
+	return err
+}
+
+func getAllAPIServersEndpoint(f *framework.Framework) ([]string, error) {
+	eps, err := f.ClientSet.CoreV1().Endpoints(metav1.NamespaceDefault).Get(context.Background(), "kubernetes", metav1.GetOptions{})
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go:938]: context.TODO()

File: vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go [hunk 5]
```diff
@@ -792,3 +786,187 @@ properties:
         type: array
         items:
           type: object`)
+
+func waitForCRDDefinitionInAllAPIServersPod(definitionName string, apiServers []string) *v1.Pod {
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go:790]: go doc

File: vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go [hunk 6]
```diff
@@ -792,3 +786,187 @@ properties:
...
+	script := `
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go:791]: go doc: summary what it does

**Comment by p0lyn0mial** [comment]
/retest

terraform . . .

File: vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go (hunk 3 above)
Comment (sttts) [reply vendor/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go:525]: setupCRDAndWaitForVersionsInSchema ?

**Comment by p0lyn0mial** [comment]
https://github.com/openshift/origin/pull/24920

**Pull request closed without merging**